admin/preprints/views.py

```diff
@@ -81,6 +87,32 @@ def get_context_data(self, **kwargs):
         kwargs.update({'message': kwargs.get('message')})
         return super(PreprintView, self).get_context_data(**kwargs)
 
+    def update_subjects_for_provider(self, request, old_provider, new_provider):
```

Comment:
If the mapping logic was shifted to the model, then this could look something like this:
```python
def update_subjects_for_provider(self, request, old_provider, new_provider):
        try:
            self.object.map_subjects_between_providers(old_provider, new_provider, Auth(request.user))
        except ValueError as e:
            messages.warning(request, e)
```

Reply:
I'd avoid using messages here or passing the request. Uhings attached to a model should rarely rely directly on the request object IMO. I think f it can't be run normally in the shell in isolation it probably shouldn't be on the model. But overall I think this should be on the model.

It also just makes this more reusable if we throw a `NodeStateError` or some internal exception and catch that in the view and insure the message is properly displayed. 

With message we can't really use this in DRF views because the can't use the messages library.

Reply:
To clarify, John, my example here is not using messages in the model.  `map_subjects_between_providers` was moved to the model, leaving `update_subjects_for_provider` in the admin view.